modules/command-engine/core/src/main/scala/surge/internal/kafka/KafkaProducerActorImpl.scala

```diff
@@ -347,86 +401,97 @@ class KafkaProducerActorImpl(
         context.become(processing(newState))
       }
     } else if (state.pendingWrites.nonEmpty) {
-      val eventMessages = state.pendingWrites.flatMap(_.publish.eventsToPublish)
-      val stateMessages = state.pendingWrites.map(_.publish.state)
-
-      val eventRecords = eventsTopicOpt
-        .map(eventsTopic =>
-          eventMessages.map { eventToPublish =>
-            // Using null here since we need to add the headers but we don't want to explicitly assign the partition
-            new ProducerRecord(eventsTopic.name, null, eventToPublish.key, eventToPublish.value, eventToPublish.headers) // scalastyle:ignore null
-          })
-        .getOrElse(Seq.empty)
-
-      val stateRecords = stateMessages.map { state =>
-        new ProducerRecord(stateTopic.name, assignedPartition.partition(), state.key, state.value, state.headers)
-      }
-      val records = eventRecords ++ stateRecords
+      // Writes grouped by requestId
+      val groupedWrites = state.groupedWrites()
```

Comment:
Hmm this groups by requestId here? So 2 different aggregates publishing would end up with 2 different request ids which would get grouped into 2 different groups? If that's the case I think this might be removing the grouping optimization we have here for transactions where we lump publish requests from multiple aggregates into one Kafka transaction to reduce overhead (one transaction per aggregate is slow and expensive).

To get around that I think instead of keeping a list of senders it could be a list of sender + requestId that we track (here are all the request ids that are in flight along with the sender who's waiting for a response)